### DISCUSSION TITLE
How to avoid OOM when restarting training in JAX/Flax?

### DISCUSSION THREAD

**Discussion by marcvanzee** [google/flax#929] [Q&A] [answered]
Suppose you train a model, then run evaluation and save a checkpoint. Then you restart the loop with training after checkpoint and the device runs into OOM. How to prevent this? Is it possible that some memory is not properly freed after evaluation?

**Comment by marcvanzee** [comment, marked as answer]
Suggestions from @avital (slightly reworded and expanded by me):

- Make sure you aren't storing the replicated parameters in the checkpoint and then re-replicating them when you restore them (assuming a multi-device setup).
- Make sure you aren't storing the checkpoint asynchronously. As long as there's a dangling Python reference to a JAX `DeviceArray`, that memory won't be de-allocated on the device. To avoid this, use `device_get` to fetch directly to host and then if you no longer have a reference to the `DeviceArray`, then it should free that from device memory.
- If your eval code is not in a separate function, any variable you assigned during eval will still have a reference forever, since Python variable scoping escapes conditionals. It could be that one of the variables is either taking a lot of memory or keeping a reference to something else that takes a lot of memory. You can avoid this by putting all the eval code in a separate function, because then references are only kept around to whatever the function returns. This is for instance the case in our [WMT example](https://github.com/google/flax/blob/master/examples/wmt/train.py), where we have functions `train_step`, `eval_step`, and `predict_step`.

**Comment by borisdayma** [comment]
I have a few instances where I run into OOM after reloading a script from a checkpoint, in non-deterministic ways (it does not happen all the time for the same script).

Is it possible it would be due to non-contiguous memory? LIke I init a model/optimizer state…, then use it to restore state with `from_bytes` (within a scoped function), eventually the temporary parameters are released but non-contiguous memory creates issues (or something else).

I was thinking to do the following right before my training steps (when all items have been correctly set/restored):
* move all items to CPU and fully release TPU's
* move them back to appropriate TPU's

Would that make sense? And if so are there already some logic/helpers in jax/flax that can make it easier?
